### PR TITLE
Return HDF5 dim. scale max size if REFERENCE_LIST atttribute missing

### LINKED ISSUE #39: Non-attached HDF5 dimension scale does not have REFERENCE_LIST attribute
A file created using the example code from the New API section of the README file causes the following error on file open:

```
ValueError                                Traceback (most recent call last)
----> 1 f = h5netcdf.File('mydata.nc', 'r')

---DIR---/h5netcdf/core.py in __init__(self, path, mode, invalid_netcdf, **kwargs)
    591         # if we actually use invalid NetCDF features.
    592         self._write_ncproperties = (invalid_netcdf is not True)
--> 593         super(File, self).__init__(self, self._h5path)
    594
    595     def _check_valid_netcdf_dtype(self, dtype, stacklevel=3):

---DIR---/h5netcdf/core.py in __init__(self, parent, name)
    242                     # variables.
    243                     self._current_dim_sizes[k] = \
--> 244                         self._determine_current_dimension_size(k, current_size)
    245
    246                     if dim_id is None:

---DIR---/h5netcdf/core.py in _determine_current_dimension_size(self, dim_name, max_size)
    284             else:  # pragma: no cover
    285                 raise ValueError(
--> 286                     "Each dimension with an actual length must have a "
    287                     "'REFERENCE_LIST' attribute.")
    288

ValueError: Each dimension with an actual length must have a 'REFERENCE_LIST' attribute.
```

This is caused by the netCDF dimension (HDF5 dimension scale) `z` in the file. The example code creates and reshapes it but never assigns to any netCDF variable. On the HDF5 side this translates to not attaching the `z` dimension scale to another HDF5 dataset. The HDF5 library creates the `REFERENCE_LIST` attribute only when a dimension scale is attached for the first time.

`h5dump` output of the `mydata.nc` file:
```
HDF5 "mydata.nc" {
GROUP "/" {
   ATTRIBUTE "_NCProperties" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "version=1|h5netcdfversion=0.5.0|hdf5libversion=1.8.18"
      }
   }
   GROUP "grouped" {
      DATASET "data" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SIMPLE { ( 10 ) / ( 10 ) }
         ATTRIBUTE "DIMENSION_LIST" {
            DATATYPE  H5T_VLEN { H5T_REFERENCE { H5T_STD_REF_OBJECT }}
            DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
            DATA {
            (0): (DATASET 5480 /grouped/y )
            }
         }
         ATTRIBUTE "foo" {
            DATATYPE  H5T_STRING {
               STRSIZE H5T_VARIABLE;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "bar"
            }
         }
      }
      DATASET "y" {
         DATATYPE  H5T_STRING {
            STRSIZE 1;
            STRPAD H5T_STR_NULLPAD;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SIMPLE { ( 10 ) / ( 10 ) }
         ATTRIBUTE "CLASS" {
            DATATYPE  H5T_STRING {
               STRSIZE 16;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "DIMENSION_SCALE"
            }
         }
         ATTRIBUTE "NAME" {
            DATATYPE  H5T_STRING {
               STRSIZE 54;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "This is a netCDF dimension but not a netCDF variable."
            }
         }
         ATTRIBUTE "REFERENCE_LIST" {
            DATATYPE  H5T_COMPOUND {
               H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
               H5T_STD_I32LE "dimension";
            }
            DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
            DATA {
            (0): {
                  DATASET 4192 /grouped/data ,
                  0
               }
            }
         }
         ATTRIBUTE "_Netcdf4Dimid" {
            DATATYPE  H5T_STD_I64LE
            DATASPACE  SCALAR
            DATA {
            (0): 1
            }
         }
      }
   }
   DATASET "hello" {
      DATATYPE  H5T_IEEE_F64LE
      DATASPACE  SIMPLE { ( 5 ) / ( 5 ) }
      ATTRIBUTE "DIMENSION_LIST" {
         DATATYPE  H5T_VLEN { H5T_REFERENCE { H5T_STD_REF_OBJECT }}
         DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
         DATA {
         (0): (DATASET 4792 /x )
         }
      }
   }
   DATASET "x" {
      DATATYPE  H5T_STRING {
         STRSIZE 1;
         STRPAD H5T_STR_NULLPAD;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 5 ) / ( 5 ) }
      ATTRIBUTE "CLASS" {
         DATATYPE  H5T_STRING {
            STRSIZE 16;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SCALAR
         DATA {
         (0): "DIMENSION_SCALE"
         }
      }
      ATTRIBUTE "NAME" {
         DATATYPE  H5T_STRING {
            STRSIZE 54;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SCALAR
         DATA {
         (0): "This is a netCDF dimension but not a netCDF variable."
         }
      }
      ATTRIBUTE "REFERENCE_LIST" {
         DATATYPE  H5T_COMPOUND {
            H5T_REFERENCE { H5T_STD_REF_OBJECT } "dataset";
            H5T_STD_I32LE "dimension";
         }
         DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
         DATA {
         (0): {
               DATASET 800 /hello ,
               0
            }
         }
      }
      ATTRIBUTE "_Netcdf4Dimid" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SCALAR
         DATA {
         (0): 0
         }
      }
   }
   DATASET "z" {
      DATATYPE  H5T_STRING {
         STRSIZE 1;
         STRPAD H5T_STR_NULLPAD;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 3 ) / ( H5S_UNLIMITED ) }
      ATTRIBUTE "CLASS" {
         DATATYPE  H5T_STRING {
            STRSIZE 16;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SCALAR
         DATA {
         (0): "DIMENSION_SCALE"
         }
      }
      ATTRIBUTE "NAME" {
         DATATYPE  H5T_STRING {
            STRSIZE 54;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SCALAR
         DATA {
         (0): "This is a netCDF dimension but not a netCDF variable."
         }
      }
      ATTRIBUTE "_Netcdf4Dimid" {
         DATATYPE  H5T_STD_I64LE
         DATASPACE  SCALAR
         DATA {
         (0): 1
         }
      }
   }
}
}
```

I am not familiar with the h5netcdf code enough but when the `REFERENCE_LIST` attribute is not present there should be no netCDF variables that depend on it so should be safe to report the maximum size of that HDF5 dimension scale as the netCDF dimension's maximum size. Something like the code below could fix this issue.

In the `h5netcdf.core.Group._determine_current_dimension_size()`:
```python
        dim_variable = _find_dim(self._h5group, dim_name)

        if "REFERENCE_LIST" not in dim_variable.attrs:
                return max_size

        root = self._h5group["/"]
```

### PR DESCRIPTION
**Pull request by ajelenak-thg**
Fixes #39

### CODE REVIEW

**Comment by shoyer** [comment]
Good catch! Can you add a test case that this error, so we can be sure this catches it?

**Comment by shoyer** [comment]
CC @krischer who wrote this logic

**Comment by ajelenak-thg** [comment]
One test added.

File: h5netcdf/core.py [hunk 1]
```diff
@@ -269,15 +269,7 @@ def _find_dim(h5group, dim):
         dim_variable = _find_dim(self._h5group, dim_name)
 
         if "REFERENCE_LIST" not in dim_variable.attrs:
-            if dim_variable.shape == (0,):
-                # NetCDF does not create the REFERENCE_LIST attribute if
-                # an unlimited dimension is of zero length. In this case it
-                # is safe to skip this dimension.
                 return max_size
```
Comment (shoyer) [inline h5netcdf/core.py]: This line needs to be dedented now.

**Comment by shoyer** [comment]
> Looks good to me assuming that no variable depends on it if REFERENCE_LIST it not present. Seems a reasonable assumption to me but I'm not familiar enough with netcdf to judge this.

REFERENCE_LIST is part of the HDF5 dimension scale API. Per the spec, it's "required if the dimension is used by any dimension." So I think this is indeed the correct logic here.

File: h5netcdf/tests/test_h5netcdf.py [hunk 2]
```diff
@@ -737,3 +737,12 @@ def test_reading_unlimited_dimensions_created_with_c_api(tmp_netcdf):
...
         f.groups['test']['dummy4'].shape == (0, 0)
+
+
+def test_reading_unused_unlimited_dimension(tmp_netcdf):
+    """Test reading a file with unused dimension of unlimited size"""
+    with h5netcdf.File(tmp_netcdf, 'w') as f:
+        f.dimensions = {'x': None}
+        f.resize_dimension('x', 5)
+
+    f = h5netcdf.File(tmp_netcdf, 'r')
```
Comment (shoyer) [inline h5netcdf/tests/test_h5netcdf.py]: Thanks for adding the test. For completeness, could you also add a check that dimensions are as expected? e.g., `assert f.dimensions == {'x': None}`.

**Comment by ajelenak-thg** [comment]
The netCDF-C library uses the HDF5 dimension scale API. The names of the HDF5 dimension scale-related attributes can be considered reserved because of the dimension scale specification. That said, these attributes are just ordinary HDF5 attributes so one could create them with the HDF5 attribute API (H5A). If any error is made in doing so (some dimension scale attributes missing, their values are not in sync) such files would violate the dimension scale specification and should be considered invalid.

In short: the proposed handling of the REFERENCE_LIST is correct.

**Comment by shoyer** [comment]
Thanks
